Src/Base/AMReX_ParallelDescriptor.H

```diff
@@ -1120,6 +1189,333 @@ Scatter(T* /*t*/, size_t /*n*/, const T1* /*t1*/, size_t /*n1*/, int /*root*/)
 }
 #endif
 
+namespace ParallelDescriptor {
+
+#ifdef AMREX_USE_MPI
+
+namespace detail {
+
+template<typename T>
+void DoAllReduce (T* r, MPI_Op op, int cnt)
+{
+#ifdef BL_LAZY
+    Lazy::EvalReduction();
+#endif
+
+    BL_ASSERT(cnt > 0);
+
+    Vector<T> recv(cnt);
+
+#if defined(BL_USE_MPI3)
+    if (doTeamReduce() > 1) {
+	Vector<T> recv_team(cnt);
+	BL_MPI_REQUIRE( MPI_Reduce(r, recv_team.dataPtr(), cnt, Mpi_typemap<T>::type(), op,
+				   0, MyTeam().get_team_comm()) );
+	if (isTeamLead()) {
+	    BL_MPI_REQUIRE( MPI_Allreduce(recv_team.dataPtr(), recv.dataPtr(), cnt,
+					  Mpi_typemap<T>::type(), op,
+					  MyTeam().get_lead_comm()) );
+	}
+	BL_MPI_REQUIRE( MPI_Bcast(recv.dataPtr(), cnt, Mpi_typemap<T>::type(),
+				  0, MyTeam().get_team_comm()) );
+    }
+    else
+#endif
+    {
+	BL_MPI_REQUIRE( MPI_Allreduce(r,
+				      recv.dataPtr(),
```

Comment:
@WeiqunZhang I am wondering if there is a reason why AMReX doesn't use `MPI_IN_PLACE` here?

Reply:
Thanks for pointing this out.  We should `MPI_IN_PLACE` here.  Will update it.